frontend/packages/dev-console/src/components/pipelineruns/logs/logs-utils.ts

```diff
@@ -0,0 +1,142 @@
+import { saveAs } from 'file-saver';
+import {
+  PodKind,
+  ContainerSpec,
+  ContainerStatus,
+  resourceURL,
+  k8sGet,
+} from '@console/internal/module/k8s';
+import { LOG_SOURCE_TERMINATED, LineBuffer } from '@console/internal/components/utils';
+import { PodModel } from '@console/internal/models';
+import { coFetchText } from '@console/internal/co-fetch';
+import { errorModal } from '@console/internal/components/modals';
+import { containerToLogSourceStatus } from '../../../utils/pipeline-utils';
+import { TaskRuns } from '../../../utils/pipeline-augment';
+
+const getSortedContainerStatus = (
+  containers: ContainerSpec[],
+  containerStatuses: ContainerStatus[],
+): ContainerStatus[] => {
+  const containerNames = containers.map((c) => c.name);
+  const sortedContainerStatus = [];
+  containerStatuses.forEach((cs) => {
+    const containerIndex = containerNames.indexOf(cs.name);
+    sortedContainerStatus[containerIndex] = cs;
+  });
+  return sortedContainerStatus;
+};
+
+export const getRenderContainers = (
+  pod: PodKind,
+): { containers: ContainerSpec[]; stillFetching: boolean } => {
+  const containers: ContainerSpec[] = pod.spec?.containers ?? [];
+  const containerStatuses: ContainerStatus[] = pod.status?.containerStatuses ?? [];
+
+  const sortedContainerStatuses = getSortedContainerStatus(containers, containerStatuses);
+
+  const firstRunningCont = sortedContainerStatuses.findIndex(
+    (container) => containerToLogSourceStatus(container) !== LOG_SOURCE_TERMINATED,
+  );
+  return {
+    containers: containers.slice(
+      0,
+      firstRunningCont === -1 ? containers.length : firstRunningCont + 1,
+    ),
+    stillFetching: firstRunningCont !== -1,
+  };
+};
+
+const getOrderedStepsFromPod = (name: string, ns: string): Promise<ContainerStatus[]> => {
+  return k8sGet(PodModel, name, ns)
+    .then((pod: PodKind) => {
+      return getSortedContainerStatus(
+        pod.spec.containers ?? [],
+        pod.status?.containerStatuses ?? [],
+      );
+    })
+    .catch((err) => {
+      errorModal({ error: err.message || 'Error downloading logs.' });
+      return [];
+    });
+};
+
+type StepsWatchUrl = {
+  [key: string]: {
+    name: string;
+    steps: { [step: string]: string };
+  };
+};
+
+export const getDownloadAllLogsCallback = (
+  sortedTaskRuns: string[],
+  taskRunFromYaml: TaskRuns,
+  namespace: string,
+  pipelineRunName: string,
+): (() => void) => {
+  const getWatchUrls = (): Promise<StepsWatchUrl> =>
+    sortedTaskRuns.reduce(async (acc, currTask) => {
+      const { pipelineTaskName, status = {} } = taskRunFromYaml[currTask];
+      const { podName } = status;
+      const steps = await getOrderedStepsFromPod(podName, namespace);
+      const allStepUrls = steps.reduce((stepUrls, currentStep) => {
+        const { name } = currentStep;
+        if (containerToLogSourceStatus(currentStep) !== LOG_SOURCE_TERMINATED) return stepUrls;
```

Comment:
We don't want to only get completed logs. If I am doing a Java build that is very noisy maven outputs and I see an error, I want to download it and look at my logs in detail - not wait for that build to finish / error out before I can get a download of the logs.

```suggestion
        if (containerToLogSourceStatus(currentStep) === LOG_SOURCE_WAITING) return stepUrls;
```
Suggestion: Avoid not started ones.